tests/blackbox.py

```diff
@@ -440,3 +477,419 @@ def dryrun_one(
+call_strat={type(approval_strat)}
+{meth_name=}
+{oc=}
+{call_cfg=}
+{is_app_create=}
+{len(self.predicates[meth_name])=}
+{stats["method_combo_count"]=}
+{stats["dryrun_count"]=}
+{stats["assertions_count"]=}
+"""
+
+        def update_stats(meth, num_preds):
+            stats[str(meth)] += num_dryruns
+            stats["method_combo_count"] += 1
+            stats["dryrun_count"] += num_dryruns
+            stats["assertions_count"] += num_dryruns * num_preds
+
+        # ---- APPROVAL PROGRAM SIMULATION ---- #
+        approval_strat = None
+        call_cfg = None
+        approve_sim = None
+        if not omit_approval_call:
+            approval_strat = ABICallStrategy(
+                json.dumps(contract.dictify()),
+                approval_args_strat_type,
+                num_dryruns=num_dryruns,
+                abi_args_mod=approval_abi_args_mod,
+            )
+            double_check_at_least_one_method = False
+            for meth_name, meth_cfg in method_configs.items():
+                sig = approval_strat.method_signature(meth_name)
+                approve_sim = Simulation(
+                    self.algod,
+                    ExecutionMode.Application,
+                    approval_teal,
+                    self.predicates[meth_name],
+                    abi_method_signature=sig,
+                    identities_teal=model_approval_teal,
+                    validation=executor_validation,
+                )
+
+                for oc_str, call_cfg in asdict(meth_cfg).items():
+                    oc = as_on_complete(oc_str)
+
+                    def simulate_approval(on_create):
+                        tp: TxParams = deepcopy(txn_params)
+                        tp.update_fields(
+                            TxParams.for_app(is_app_create=on_create, on_complete=oc)
+                        )
+                        sim_results = approve_sim.run_and_assert(
+                            approval_strat, txn_params=tp, msg=msg4simulate()
+                        )
+                        assert sim_results.succeeded
+                        self.auto_path_results[meth_name][(on_create, oc)] = sim_results
+                        update_stats(meth_name, len(self.predicates[meth_name]))
+
+                    # weird walrus is_app_create := ... to fill closure of msg4simulate()
+                    if call_cfg & CallConfig.CALL:
+                        double_check_at_least_one_method = True
+                        simulate_approval(is_app_create := False)
+
+                    if call_cfg & CallConfig.CREATE:
+                        double_check_at_least_one_method = True
+                        simulate_approval(is_app_create := True)
+            assert double_check_at_least_one_method, "no method was simulated"
+
+        # ---- CLEAR PROGRAM SIMULATION ---- #
+        clear_sim: Simulation | None = None
+        if not omit_clear_call:
+            if isinstance(clear_args_strat_type_or_inputs, list):
```

Comment:
this line looks a bit off: `clear_args_strat_type_or_inputs` is a passed in argument of type `Type[ABIStrategy]`, but how is it able to test against `isinstance(..., list)`?

Reply:
The type's are wacky and you are better than mypy!!!! Let's see how painful it is to get mypy to detect it.

Reply:
Yea, I think it is somehow contradicting in a way that:
- if we stick with `clear_args_strat_type_or_inputs` being a list-like instance, then when it get passed into `_validate_simulation`, with `omit_clear_call = False`, it will fail the assert.

Reply:
I didn't see your second comment till now. I'll investigate further.

Reply:
oh I see, my bad. I didn't see the union type breakline down there. the change now looks good.

Reply:
> I didn't see your second comment till now. I'll investigate further.

no, my bad, I didn't see the union type breakline, now I passed again and noticed the union type of `clear_args_strat_type_or_inputs`